### PR TITLE
[build][pytorch] pt1.10 add openssh support

### PR DESCRIPTION
**Pull request by Qingzi-Lan**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu [hunk 1]
```diff
@@ -72,6 +72,27 @@ ENV PATH="$PATH:/home/.openmpi/bin"
 ENV LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/home/.openmpi/lib/"
 RUN ompi_info --parsable --all | grep mpi_built_with_cuda_support:value
 
+# Install OpenSSH for MPI to communicate between containers, allow OpenSSH to talk to containers without asking for confirmation
+RUN apt-get update \
+ && apt-get install -y  --allow-downgrades --allow-change-held-packages --no-install-recommends \
```
Comment (tejaschumbalkar) [inline pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu:77]: is this line required? Seems redundant

Comment (Qingzi-Lan) [reply pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu:77]: I directly use the install commands in GPU training dockerfile [here](https://github.com/aws/deep-learning-containers/blob/master/pytorch/training/docker/1.8/py3/cu111/Dockerfile.gpu#L287-L305). prefer to leave it as is to avoid unexpected behaviors.

**Review by spidyDev — COMMENTED** [review]
LGTM!

File: pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu (hunk 1 above)
Comment (saimidu) [reply pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu:77]: This change seems to have originated here: https://github.com/aws/deep-learning-containers/pull/356#discussion_r461830982

I recommend trying to run one execution of tests to see whether it works without this line, because it sees like it should be completely redundant.

Comment (Qingzi-Lan) [reply pytorch/training/docker/1.10/py3/Dockerfile.e3.cpu]: rerun the build and test for e3 cpu training without that line. tests pass

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 2]
```diff
@@ -125,6 +125,19 @@ def test_pytorch_gloo(pytorch_training, ec2_connection, gpu_only, py3_only, ec2_
...
 
+@pytest.mark.integration("gloo")
+@pytest.mark.model("resnet18")
+@pytest.mark.parametrize("ec2_instance_type", PT_EC2_CPU_INSTANCE_TYPE, indirect=True)
+def test_pytorch_gloo_cpu(pytorch_training, ec2_connection, cpu_only, py3_only, ec2_instance_type):
+    """
+    Tests gloo backend
+    """
+    if get_processor_from_image_uri(pytorch_training):
+        pytest.skip(f"{__name__} is for CPU image only")
```
Comment (saimidu) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:136]: The returned `processor` from the `get_processor_from_image_uri` can be one of `["eia", "neuron", "cpu", "gpu", "hpu"]`.

If the function fails to find a processor, it throws a `RuntimeError`. Writing the skip condition this way would cause this test to be skipped in all executions. You would want to compare the output of the `get_processor_from_image_uri` to `"cpu"`.

You don't need to do this, though, because `cpu_only` in the test function fixtures would cause `pytorch_training` to be parametrized in such a way that only CPU DLC images are assigned to it.

Could you please change this so that the pytest.skip and its associated if-condition are removed?

File: test/dlc_tests/container_tests/bin/pytorch_tests/testPyTorchGlooCpu [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
```
Comment (saimidu) [inline test/dlc_tests/container_tests/bin/pytorch_tests/testPyTorchGlooCpu:1]: Please modify the pre-existing `testPyTorchGloo` script by adding conditions to determine whether the test should be run with CPU args or with GPU args, rather than adding a whole new wrapper script around `dlbs/python/pytorch_benchmarks/benchmarks.py` just to test with CPU.

File: test/test_utils/eks.py [hunk 4]
```diff
@@ -273,7 +273,7 @@ def eks_forward_port_between_host_and_container(selector_name, host_port, contai
     )
 
 
-@retry(stop_max_attempt_number=20, wait_fixed=30000, retry_on_exception=retry_if_value_error)
+@retry(stop_max_attempt_number=30, wait_fixed=30000, retry_on_exception=retry_if_value_error)
```
[the remark this answers is not in the document]
Comment (Qingzi-Lan) [reply test/test_utils/eks.py:276]: the worker node was not ready in 20 retrys. it will cause the test fail. increase to 30 to get more time to initialize a worker node.

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py (hunk 2 above)
Comment (saimidu) [reply test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:136]: Note: If you unify the `testPyTorchGlooCpu` script with the `testPyTorchGloo` script, then you wouldn't need another test function specifically for CPU DLCs.

File: test/test_utils/eks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (saimidu) [reply test/test_utils/eks.py:276]: The worker node initialization took more than 10 minutes? This may be something that warrants a closer look. Sure, let's increase the number of tries to 30, but let's also dive deeper to root-cause this.

**Review by saimidu — DISMISSED** [review]
Approved. Please revert all dev config and buildspec changes, and let's merge this.

File: test/dlc_tests/container_tests/bin/testSmprofiler [hunk 5]
```diff
@@ -119,7 +119,7 @@ if [[ ${FRAMEWORK} == "pytorch" ]]; then
         echo "SMProfiler pytorch [True-False] zcc integration test failed"
         exit $exit_code
     fi
-    python -m pytest --assert=plain pytest "tests/zero_code_change/test_pytorch_integration.py::test_pytorch[False-False]"
+    python -m pytest --assert=plain "tests/zero_code_change/test_pytorch_integration.py::test_pytorch[False-False]"
```
[the remark this answers is not in the document]
Comment (saimidu) [reply test/dlc_tests/container_tests/bin/testSmprofiler:122]: Thanks. Context: https://github.com/aws/deep-learning-containers/pull/1709/commits/36357c7974a33fdecf2e611d66390b6bd847dab7

**Review by saimidu — APPROVED** [review]
